modules/mbedtls/CMakeLists.txt

```diff
@@ -1,71 +1,210 @@
+#
+# Copyright (c) 2023 Linaro Limited
+#
+# SPDX-License-Identifier: Apache-2.0
+#
 if(CONFIG_MBEDTLS)
-zephyr_interface_library_named(mbedTLS)
 
-if(CONFIG_MBEDTLS_BUILTIN)
-  target_compile_definitions(mbedTLS INTERFACE
-    MBEDTLS_CONFIG_FILE="${CONFIG_MBEDTLS_CFG_FILE}"
-  )
+  if(CONFIG_MBEDTLS_BUILTIN)
+    if(CONFIG_MBEDTLS_ZEPHYR_ENTROPY AND NOT CONFIG_ENTROPY_HAS_DRIVER)
+    message(WARNING "No entropy device on the system, using fake entropy source!")
+    endif()
 
-  target_include_directories(mbedTLS INTERFACE
-    ${ZEPHYR_CURRENT_MODULE_DIR}/include
-    # mbedTLS v3.1.0 has psa_crypto_cipher.c including an internal header using <>.
-    # This line ensures the header can be found in Zephyr.
-    # When updating to mbedTLS > v3.1.0, please check if this has been corrected,
-    # and if so remove this include path.
-    ${ZEPHYR_CURRENT_MODULE_DIR}/library
-    configs
-    include
-  )
+    # Library for mbed TLS includes and configurations
+    add_library(mbedTLS INTERFACE)
```

Comment:
I want to expose the mbedTLS headers whoever target_link_libraries with mbedTLS so created as the interface libraries, however, the zephyr extension zephyr_library_include_directories same function calls the target_include_directories indirectly with PRIVATE to prevent exposing the headers and same issue with zephyr_library_compile_definitions.

Because of that using the direct target_include_directories and target_compile_definitions CMake calls, so called CMake function APIs for creating the mbedTLS headers interface library rather than using zephyr extension functions.

Reply:
> What is not working when using the zephyr functions?

It sounds like due to the `PRIVATE` flag in the zephyr extensions, the header files are not being exported, but @rajkan01 can hopefully clarify further.

Reply:
yes, it is a problem with the PRIVATE flag in the zephyr extensions.

Reply:
Can you give an example because I switched in this PR to use the old method and build smp_svr with sha256 support and it built fine?

Reply:
You might want to try the `psa_crypto` sample in this PR, since that brings out the problems we're trying to solve with MbedTLS. Some variation of this (QEMU used below):

```bash
$ west build -p auto -b mps2_an521_ns path/psa_crypto -t run
```

Reply:
> You might want to try the psa_crypto sample in this PR, since that brings out the problems we're trying to solve with MbedTLS. Some variation of this (QEMU used below):

Tried that using the old `zephyr_interface_library_named(mbedTLS)` style with this PR and it builds fine.

> I think no, if the particular application or other libraries depend on mbedTLS, they should explicitly link libraries with mbedTLS as other does in the Zephyr source tree.

But that's no different than it was and using the zephyr_ functions is no different in that regard?

> Another problem, I could not find a way to overcome the issue of automatically adding header inclusion using Zephyr CMake, if you have suggestions to fix this, that would be good.

As in adjusting the include paths for the zephyr application as a whole?